packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -0,0 +1,581 @@
+/// of the widgets in the segmented control.
+///
+/// When the state of the segmented control changes, the widget calls the
+/// [onValueChanged] callback. The key associated with the newly selected
+/// widget is returned in the [onValueChanged] callback. Most widgets that
+/// use a segmented control will listen for the [onValueChanged] callback
+/// and rebuild the segmented control with a new [groupValue] to update which
+/// option is currently selected.
+///
+/// The height of the segmented control is determined by the height of the
+/// tallest widget provided as a value in the [LinkedHashMap] of [children].
+/// The width of the segmented control is determined by the horizontal
+/// constraints on its parent. The available horizontal space is divided by
+/// the number of provided [children] to determine the width of each widget.
+/// The selection area for each of the widgets in the [LinkedHashMap] of
+/// [children] will then be expanded to fill the calculated space, so each
+/// widget will appear to have the same dimensions.
+///
+/// See also:
+///
+///  * <https://developer.apple.com/design/human-interface-guidelines/ios/controls/segmented-controls/>
+class SegmentedControl<T> extends StatefulWidget {
+  /// Creates an iOS-style segmented control bar.
+  ///
+  /// The [children] and [onValueChanged] arguments must not be null. Further,
+  /// the length of the [children] list must be greater than or equal to two.
+  ///
+  /// Each widget value in the map of [children] must have an associated key
+  /// that uniquely identifies this widget. This key is what will be returned
+  /// in the [onValueChanged] callback when a new value from the [children] map
+  /// is selected. Most widgets that use a segmented control will listen for
+  /// the [onValueChanged] callback and rebuild the segmented control with
+  /// the returned key as the new [groupValue] to update the currently selected
+  /// option.
+  ///
+  /// The [groupValue] must be one of the keys in the [children] map.
+  /// The [groupValue] is the currently selected value for the segmented control.
+  /// If no [groupValue] is provided, or the [groupValue] is null, no widget will
+  /// appear as selected.
+  SegmentedControl({
+    Key key,
+    @required this.children,
+    @required this.onValueChanged,
+    this.groupValue,
+  })  : assert(children != null),
+        assert(children.length >= 2),
+        assert(onValueChanged != null),
+        assert(groupValue == null || children.keys.any((T child) => child == groupValue)),
+        // in case the default implementation of Map changes
+        assert(children is LinkedHashMap),
+        super(key: key);
+
+  /// The identifying keys and corresponding widget values in the
+  /// segmented control.
+  ///
+  /// This attribute must not be null. The length of the list must be greater
+  /// than or equal to two.
+  final Map<T, Widget> children;
```

Comment:
this should be LinkedHashMap, then you can drop the assert.

Reply:
oh never mind, `const { }` doesn't give you a LinkedHashMap. Just remove the assert above and leave this as a Map.

Reply:
Why remove the assert?  Since the implementation depends on it being an ordered Map, it seems like you want to make sure that it really is one, and if the underlying representation were to change, you want it to fail.

Or are you saying that this won't work because it'll assert with a `const <T, Widget>{}`?  Is that still ordered (I assume so)?

Reply:
const maps are some other internal class that doesn't implement LinkedHashMap. I'm pretty sure they're still ordered though.